[UniFi] Initial contribution

mgbowman opened the pull request:
This is the initial contribution of the UniFi binding that has been over a year in the making.

There's been an active discussion and testing effort over at [Ubiquiti Unifi Binding Feature Discussion](https://community.openhab.org/t/ubiquiti-unifi-binding-feature-discussion/14520)

The latest successful build artifacts can be download from my own personal Jenkins instance:

- [org.openhab.binding.unifi-2.3.0-SNAPSHOT.jar](https://jenkins.otr.mx/job/openhab2-unifi-binding/lastSuccessfulBuild/artifact/addons/binding/org.openhab.binding.unifi/target/org.openhab.binding.unifi-2.3.0-SNAPSHOT.jar)
- [Static Code Analysis Report](https://jenkins.otr.mx/job/openhab2-unifi-binding/lastSuccessfulBuild/artifact/addons/binding/org.openhab.binding.unifi/target/code-analysis/report.html)

Looking forward to this making into master such that others can contribute.

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific addon: Mention the addon shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the addon README?
- Does your contribution follow the coding guidelines:
  http://docs.openhab.org/developers/development/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  http://docs.openhab.org/developers/development/bindings.html#static-code-analysis
- Did you sign-off your work:
  http://docs.openhab.org/developers/contributing/contributing#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It's a good practice to add an URL to your built JAR in this Pull Request description,
so it's easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it's reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

martinvw commented:
Hooray!!!

martinvw commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -1,12 +1,12 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
 
-    Copyright (c) 2010-2018 by the respective copyright holders.
```
Can you leave this alone and add your binding in this file

mgbowman replied:
I think `mvn license:format` got this file 🤦‍♂️

mgbowman commented:
@martinvw I fixed the feature.xml mishap. Anything else you see wrong?

martinvw commented:
A full review will take me some more time, let me try to do it tonight :-)

martinvw commented:
@wborn your first public action, welcome! Feel free to do the review as well. I'm personally really looking forward to this binding I already run it for over 8 months.

wborn commented:
Thanks @martinvw! I hope this binding also inspires someone to make something similar for AsusWRT. :-)

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,137 @@
+				<default>unifi</default>
+				<context>network-address</context>
+			</parameter>
+			<parameter name="port" type="integer" max="65535" min="1" required="true">
```
can be required=false as you have a default

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,137 @@
+				<label>Site</label>
+				<description>The UniFi site where the client should be found (optional)</description>
+			</parameter>
+			<parameter name="contactType" type="text" required="false">
```
Not really clear to me what this parameter is good for.

mgbowman replied:
I added this so you could change the type of Contact - Normally Open vs Normally Closed. If I change the online channel to a Switch type, this parameter will go away.

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,137 @@
+				<description>The refresh interval in seconds to poll the UniFi controller</description>
+				<default>10</default>
+			</parameter>
+			<parameter name="considerHome" type="integer" required="false">
```
Shouldn't this rather be a parameter on the client?

mgbowman replied:
This would be more flexible as you could specify how quickly the binding would consider a client as away on a per-client basis.

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,137 @@
+	</thing-type>
+	
+	<channel-type id="online">
+		<item-type>Contact</item-type>
```
Contact is pretty unusual here - we usually use Switch, see e.g. the Network binding, which has a similar feature.

mgbowman replied:
We talked a little about this in the discussion thread and I opted for Contact since the HOME vs AWAY state is read only. I can change it back to Switch if necessary.

kaikreuzer replied:
Read-only shouldn't be an argument as you can declare that for switches as well - so yes, I would very much prefer to have this changed to Switch.

dimalo replied:
Actually the FritzBox TR64 OH 1.x binding I use at the moment also uses Contacts for online detection...

kaikreuzer replied:
Should be changed as well then - latest when being refactored to a 2.x binding...

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,137 @@
+	<channel-type id="site">
+		<item-type>String</item-type>
+		<label>Site Name</label>
+		<description>UniFi Site the client is associated with</description>
```
What is the Site and can that change dynamically? If not, that should rather be an (informative) Thing property.

mhilbush replied:
@kaikreuzer I'll chime in on this one since I was the one who asked for this feature. ;-)

A Unifi controller can manage multiple sites (a site is a place where you have Unifi switches, APs, etc. deployed). In the case where you have a controller managing more than one site (e.g. Home A and Home B), and you also have OH instances in each site, you need a way to tell the binding the site for which you want to know if a client (e.g. phone) is present. Without the ability to specify the site on the client thing, the binding would report the client present in Home A when the device is physically located in Home B (and vice versa).

Additional discussion here.
https://community.openhab.org/t/ubiquiti-unifi-binding-feature-discussion/14520/63

mgbowman replied:
Following up on this, yes the site can change dynamically. I have one instance of the UniFi controller that manages my home and office. As I move from my home to my office, my phone's site changes from "Home" to "Work". If my phone goes offline and the last site was "Work", I'm on my way home.

kaikreuzer replied:
ok, thanks for the explanation!

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,31 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-ClassPath: .
+Import-Package: 
+ com.fasterxml.jackson.annotation,
```
Could you use GSON instead? See https://www.eclipse.org/smarthome/documentation/development/bindings/dependencies.html#for-json-processing

mgbowman replied:
I can, but this will require a bit of rework (and testing). I'll fork another branch for this and seek help from the community to test it.

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/README.md:
```diff
@@ -0,0 +1,127 @@
+
+## Discovery
+
+Discovery is not currently supported.
```
-> currently not

mgbowman replied:
My kind of change request 😉

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+
+    @Override
+    public void dispose() {
+        cancelRefreshJob();
```
Should we close / null the controller here?

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+ *
+ * @author Matthew Bowman - Initial contribution
+ */
+public class UniFiControllerHandler extends BaseBridgeHandler implements Runnable {
```
Please do no implement the runnable for this class

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    @Override
+    public void run() {
```
@wborn what is in your opinion the best thing to do here, implement runnable, just call this method, or directly call refresh and trust the logging from the framework and accept the fact that is might break the update if some uncatched exception occurs

mgbowman replied:
While some users were testing the binding, we had some NPEs being thrown (because of some weird data returned by the UniFi controller) that weren't being logged and the binding would just hang. After I figured out the subsequent calls to refresh were being suppressed because of the uncaught exceptions, I added the try/catch logic here.

I'm guessing things have changed with the logging, but how do you guys recommend I go about gracefully handling uncaught exceptions from a "refreshJob" point of view? If I do what you're suggesting, any error basically puts the binding in a frozen state and there will be no more updates until either the binding is reconfigured or OH is restarted.

wborn replied:
> that weren't being logged

mgbowman replied:
That makes sense.

What should I do about the refresh job? If I do it the way you suggest, any unhandled error will cause the binding to freeze.

wborn replied:
After some testing it seems that the `WrappedScheduledExecutorService` only logs the exception and after that further scheduled jobs no longer execute. :-(

I've created a simple exception binding with a thing handler that [schedules a job](https://github.com/wborn/openhab2-addons/blob/f2291e98a3d3832b6a2a4affa6903475ffb316df/addons/binding/org.openhab.binding.exception/src/main/java/org/openhab/binding/exception/handler/ExceptionThingHandler.java#L56) to throw a certain exception type. After the first execution it logs the error but thereafter the scheduled job no longer runs:

```
21:45:26.215 [WARN ] [ommon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception
java.lang.NullPointerException: NullPointerException Message
	at org.openhab.binding.exception.handler.ExceptionThingHandler.getConfiguredException(ExceptionThingHandler.java:93) ~[?:?]
	at org.openhab.binding.exception.handler.ExceptionThingHandler.runScheduledCommand(ExceptionThingHandler.java:72) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

So for now I agree you can better implement the try/catch yourself. 

@Hilbrand is this a bug in the WrappedScheduledExecutorService? If this is not a bug then I think the JavaDoc of that class should be updated so it is more clear what kind of burden is exactly removed for the user.

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+        String statusDescription = getThing().getStatusInfo().getDescription();
+
+        if (controller == null) {
+            // mgb: this should never happen
```
Given the message below, I think that we can live without this line of comment

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+            logger.warn("UniFi Controller is NULL - this should never happen!");
+        } else {
+            // mgb: we're offline, we need to try to login
+            if (status == INITIALIZING || (status == OFFLINE && statusDetail == COMMUNICATION_ERROR)) {
```
@kaikreuzer I never saw such logic before in a binding is this expected for you?

mgbowman replied:
After looking over this again and again, I kept saying to myself "what were you thinking here?!?". It finally dawned on me why I did it this way.

I was originally checking whether the controller credentials were valid in the `initialize()` method, but it occurred to me the controller could be offline / unreachable and I would wrongly report this as an invalid configuration (invalid credentials).

I'm going to think about how to clean this up to make it clearer.

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+                    // mgb: ... then refresh all the clients
+                    for (Thing thing : getThing().getThings()) {
+                        ThingHandler handler = thing.getHandler();
+                        if (handler instanceof UniFiClientHandler) {
```
Can it be something else, should we do some logging in that case?

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+                        }
+                    }
+                } catch (UniFiException e) {
+                    logger.warn("Error refreshing the UniFi Controller {} - {}", getThing().getUID(), e.getMessage());
```
Do we want to ignore the stacktrace, can we still find out what went wrong?

martinvw replied:
See:

> Do we want to ignore the stacktrace, can we still find out what went wrong?

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        // mgb: update the status if it's changed
+        if (status != getThing().getStatus()) {
```
We might have a different OFFLINE cause now. why should we check?

mgbowman replied:
I added the check because without the check, the log gets filled with `ThingStatusInfoEvent` messages:

```
2018-04-12 21:28:48.487 [INFO ] [ome.event.ThingStatusInfoEvent:47   ] - 'unifi:controller:home' updated: ONLINE
2018-04-12 21:28:58.538 [INFO ] [ome.event.ThingStatusInfoEvent:47   ] - 'unifi:controller:home' updated: ONLINE
2018-04-12 21:29:08.596 [INFO ] [ome.event.ThingStatusInfoEvent:47   ] - 'unifi:controller:home' updated: ONLINE
2018-04-12 21:29:18.658 [INFO ] [ome.event.ThingStatusInfoEvent:47   ] - 'unifi:controller:home' updated: ONLINE
```

martinvw replied:
@htreu is this expected I thought we did not log repeated events?

martinvw replied:
Of course that is welcome, note that it will get you in ESH (Eclipse smart home) project

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+        return result;
+    }
+
+    // Public API functions
```
I would expect those below the private methods (on which they depend)

martinvw replied:
Ahum, I hope you read this as above :-S

mgbowman replied:
So you mean above?

Just to be clear...
```
// Public API

public foo() {
     bar();
}

// Private API

private bar() {
    logger.debug("martinv needs to make up his mind");
}
```

🙊

martinvw replied:
😁

Yes first public methods, but they don't have to be in blocks. Having a public method immediately followed by its private helper methods is also fine.

But if you use blocks, public should on top after your constructors.

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+            SSLContext sc = SSLContext.getInstance("SSL");
+            sc.init(null, new TrustManager[] { new UniFiTrustManager() }, new SecureRandom());
+            sslSocketFactory = sc.getSocketFactory();
+        } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

mgbowman replied:
Honestly, I couldn't think of any "good" way to handle this as the UniFi controller uses a self-signed SSL certificate and if any of this logic fails (installing the blind trust manager), the binding won't function at all. 

I decided the best course of action was to catch Exception and rethrow a UniFiException to signal the issue in the PaperUI as a `ThingStatusDetail.CONFIGURATION_ERROR`.

martinvw replied:
But what exceptions can come from the block given you implemented some accept all logic for the certificates

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+                response = objectMapper.readValue(content, responseType);
+            }
+        } catch (SocketException | SocketTimeoutException e) {
+            throw new UniFiException("Could not communicate with the UniFi Controller");
```
Should we preserve the stacktrace here?

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+        } catch (SocketException | SocketTimeoutException e) {
+            throw new UniFiException("Could not communicate with the UniFi Controller");
+        } catch (Exception e) {
+            logger.error("Error GET'ing URL: {}", url, e);
```
In general its a best practice to either log or rethrow, by doing both you might end up with twice the same error (or stacktrace) in your log.

martinvw replied:
See above:

> In general its a best practice to either log or rethrow, by doing both you might end up with twice the same error (or stacktrace) in your log.

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+            }
+        } catch (SocketException | SocketTimeoutException e) {
+            throw new UniFiException("Could not communicate with the UniFi Controller");
+        } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+    public boolean login() {
+        boolean success = false;
+        // login
+        Map<String, String> params = new HashMap<String, String>();
```
Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

mgbowman replied:
I'll search the codebase for all instances of this and fix them all.

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+        } catch (SocketException | SocketTimeoutException e) {
+            logger.warn("Could not connect to the UniFi Controller");
+            success = false;
+        } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw replied:
> Do you really need to catch the Exception-class or are there more specific exceptions you could catch

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+            String response = get(getLoginUrl(), params);
+            success = (response != null);
+        } catch (SocketException | SocketTimeoutException e) {
+            logger.warn("Could not connect to the UniFi Controller");
```
Should we log the exception (message) here as well?

martinvw replied:
See above:

> Should we log the exception (message) here as well?

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+
+    private Map<String, UniFiSite> getSites() throws UniFiException {
+        UniFiSiteResponse response = get(getSitesUrl(), UniFiSiteResponse.class);
+        Map<String, UniFiSite> siteMap = new HashMap<String, UniFiSite>();
```
Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

Please check all

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiClient.java:
```diff
@@ -0,0 +1,92 @@
+
+    protected String hostname;
+
+    @JsonProperty("site_id")
```
You can get rid of this problem in one go:

https://stackoverflow.com/a/11810183/3656774

wborn commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiControllerConfig.java:
```diff
@@ -0,0 +1,65 @@
+public class UniFiControllerConfig {
+
+    private String host;
+    private BigDecimal port;
```
When you replace `BigDecimal` with `int` you don't have to call `.intValue()` everywhere which will make the code more readable.

wborn commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiControllerConfig.java:
```diff
@@ -0,0 +1,65 @@
+    private BigDecimal port;
+    private String username;
+    private String password;
+    private BigDecimal refresh;
```
When you replace `BigDecimal` with `int` you will not have to call `.intValue()` everywhere which will make the code more readable.

martinvw replied:
@wborn are you sure, Afaik there only limited types supported by the configuration classes used via the ESH getConfiguration.as(...class)

wborn replied:
Yes the Configuration class is using [reflection logic](https://github.com/eclipse/smarthome/blob/6852de6b356f76e7d832e629034e7ea2b0d3c349/bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/Configuration.java#L90) for this. I've used it myself this way e.g. in the [PlugwiseSenseConfig](https://github.com/openhab/openhab2-addons/blob/d3e80386d3b67a7c7c47b74fbeac1d0ca385a369/addons/binding/org.openhab.binding.plugwise/src/main/java/org/openhab/binding/plugwise/internal/config/PlugwiseSenseConfig.java#L31) class.

martinvw replied:
Great, then there is some room for improvements in some bindings 😃

wborn commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiControllerConfig.java:
```diff
@@ -0,0 +1,65 @@
+    private String username;
+    private String password;
+    private BigDecimal refresh;
+    private BigDecimal considerHome;
```
When you replace `BigDecimal` with `int` you don't have to call `.intValue()` everywhere which will make the code more readable.

wborn commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiException.java:
```diff
@@ -0,0 +1,38 @@
+
+    /**
+     *
+     */
```
Please remove this empty comment.

wborn commented on addons/binding/org.openhab.binding.unifi/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,31 @@
+ com.fasterxml.jackson.core,
+ com.fasterxml.jackson.databind,
+ com.fasterxml.jackson.databind.annotation,
+ com.google.common.collect,
```
You can remove the dependency on `com.google.common.collect` when you replace the set logic with Java 8 code like:

```
    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Stream
            .of(UniFiBindingConstants.THING_TYPE_CONTROLLER, UniFiBindingConstants.THING_TYPE_CLIENT)
            .collect(Collectors.toSet());
```

wborn commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+        try {
+            controller = new UniFiController(config);
+        } catch (UniFiException e) {
+            logger.error("Error configuring the UniFi Controller: {}", e.getMessage());
```
The stacktrace is usually only logged when debug logging is enabled.

mgbowman replied:
If there's a `UniFiException` caught at this stage, the message is a descriptive reason that is logged both in the console and displayed in the PaperUI (via the next line of code).

It's meant to catch configuration errors where the user provider invalid parameters (wrong username, wrong password, etc...) and put the binding into a `ThingStatusDetail.CONFIGURATION_ERROR` state.

martinvw replied:
But I saw some places where (generic) exceptions were converted to the UnifiException 
, maybe you could make a separate subclass or rename the UnifiException to UnifiConfigurationException

wborn commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+
+        if (!config.isValid()) {
+            throw new UniFiException("Invalid configuration");
+        } else if (config.getConsiderHome().compareTo(config.getRefresh()) <= 0) {
```
See comment in `UniFiControllerConfig`.
This expression will be simplified when the `BigDecimal` configuration values are replaced with `int` values. :-)

wborn commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,403 @@
+        try {
+            get(getLogoutUrl());
+        } catch (Exception e) {
+            // nop
```
Logging and narrowing down the caught exception to `IOException | URISyntaxException`  would be better.

martinvw replied:
> Logging and narrowing down the caught exception to IOException | URISyntaxException would be better.

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiClientHandler.java:
```diff
@@ -0,0 +1,215 @@
+            logger.debug("Refreshing channel = {}", channelUID);
+
+            UniFiController controller = ((UniFiControllerHandler) getBridge().getHandler()).getController();
+            UniFiWirelessClient client = controller.getWirelessClient(config.getMac(), config.getSite());
```
Yes everything is cached.

The data from the controller is collected via 4 different sequential API calls:
 1. Grab the Sites (UniFi Organization)
 1. Grab the Devices (Access Points, Switches)
 1. Grab the Clients (Wired / Wireless Clients)
 1. Grab the Insights (Statistical Data)

The JSON data returned by the controller contains the relevant data relationships, but only based on IDs. As I retrieve and parse the data, I lookup the respective objects by IDs and relate them on a POJO level for an easy to use / navigate Java data model. This whole process happens every refresh interval.

These `.getWirelessClient(..)` calls are simply hitting the cache of the latest data that was returned by the controller.

martinvw replied:
Great

mgbowman commented:
@martinvw @wborn 

I made most of the minor changes you guys requested and committed them individually to my local repo. Do you want me to squash them before pushing or would you rather me leave them as individual commits to make review easier?

mgbowman commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,137 @@
+			<channel id="site" typeId="site" />
+			<channel id="ap" typeId="ap" />
+			<channel id="essid" typeId="essid" />
+			<channel id="rssi" typeId="rssi" />
```
Would this mean I need to map the RSSI to a numeric range of 0..4 as stated in the docs?

https://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#channels

martinvw replied:
If that is possible that would be great imho

mgbowman replied:
I researched this and I couldn't find a clear answer to what the controller's (min,max) range is.

From wikipedia:

> Vendors and chipset makers provide their own accuracy, granularity, and range for the actual power (measured as milliwatts or decibels) and their range of RSSI values (from 0 to RSSI maximum).

Until I can find a definitive range, I don't think there's a reliable way to map this into a 0..4 range. 

Is it ok if we table this for a future release?

martinvw replied:
👍

mgbowman commented on addons/binding/org.openhab.binding.unifi/OSGI-INF/UniFiHandlerFactory.xml:
```diff
@@ -0,0 +1,20 @@
+	http://www.eclipse.org/legal/epl-v10.html
+
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true" name="binding.unifi">
```
I think you're saying to delete this file and create `OSGI-INF/.gitignore` with
```
/*.xml
```

Correct?

I'm assuming the framework handles generating the XML at runtime.

martinvw replied:
Close it is handled at compile time. And it also requires you to add the correct annotations on the Java code.

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerHandler.java:
```diff
@@ -0,0 +1,172 @@
+    @Override
+    public void initialize() {
+        Version version = bundleContext.getBundle().getVersion();
+        logger.info("UniFi Binding v{}", version);
```
I can make this debug. I put this in here during development so people could easily see which version they were running as the binding was changing very frequently in the beginning.

martinvw replied:
There are also other ways to see this binding for example the karaf console, so maybe it shouldn't even be part of the binding @kaikreuzer wdyt?

martinvw commented:
I personally prefer to have separate commits, as long as you sign them off we just squash them for ou in the end.

mgbowman commented:
@martinvw I just pushed the commits - have a look.

I'll start working on the more complex changes soon.

martinvw commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,137 @@
+				<description>The refresh interval in seconds to poll the UniFi controller</description>
+				<default>10</default>
+			</parameter>
+			<parameter name="considerHome" type="integer" required="false">
```
> I'm not sure whether this should be part of the binding or is a generic concept, @kaikreuzer wdyt?

@kaikreuzer, see my question above 🙂

mgbowman replied:
While I do believe this is probably better suited as a generic concept in the OH framework, I feel this parameter needs to stay until it can be refactored to a higher level. The way the UniFi controller caches the connected clients, we really need a way to force the binding to mark a client as offline even though they still appear in the UniFi controller. Without it, we get long delays in marking a client offline, which is undesirable and leads users to think the binding is not working properly.

I will happily refactor this if the OH core ever provides a generic "considerHome" parameter.

martinvw replied:
I believe I had something else in my mind, this timeout parameter does not seem to be a problem, sorry

martinvw commented:
@mgbowman just a sanity check, how are you proceeding here? AFAIK I'm still waiting for you 🙂 

> I'll start working on the more complex changes soon.

mgbowman commented:
@martinvw it's been busy lately and with the holidays I didn't have much time. I'm going to do my best to try and wrap these changes up this week so I can move on to the Spotify binding.

I'll keep you posted.

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiClientHandler.java:
```diff
@@ -0,0 +1,215 @@
+            return;
+        }
+
+        if (bridge.getStatus() == OFFLINE) {
```
I'm not sure what you mean. At this point in the code, the client is being configured and if the controller (bridge) isn't online yet, does it make sense for the client to be marked as online?

I did it this way because I saw `ThingStatusDetail.BRIDGE_OFFLINE` and thought this was exactly its use case.

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiClientHandler.java:
```diff
@@ -0,0 +1,215 @@
+
+                // :essid
+                case CHANNEL_ESSID:
+                    if (clientOnline && client != null && StringUtils.isNotBlank(client.getEssid())) {
```
`clientOnline` is only true if there is a client _and_ its `considerHome` interval has not passed (based on the client's `lastSeen` value).

When we force the binding to mark a client as away (via `considerHome`), there's a window where the controller still returns valid data. I didn't think it made sense to keep updating the client channels if the client is away.

Without this check, when a client is marked as offline, all of its channels (with the exception of `lastSeen`) would still get updates posted to the bus.

Maybe I should rename this to `clientHome` to make it more clear, wdyt?

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiClientHandler.java:
```diff
@@ -0,0 +1,215 @@
+
+                // :rssi
+                case CHANNEL_RSSI:
+                    if (clientOnline && client != null && client.getRssi() != null) {
```
See above

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiClientHandler.java:
```diff
@@ -0,0 +1,215 @@
+
+                // :uptime
+                case CHANNEL_UPTIME:
+                    if (clientOnline && client != null && client.getUptime() != null) {
```
See above

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiClientHandler.java:
```diff
@@ -0,0 +1,215 @@
+                logger.debug("Could not find a matching client: mac = {}, site = {}", config.getMac(),
+                        config.getSite());
+            }
+            boolean clientOnline = controller.isWirelessClientOnline(client);
```
See my comments below.

I did it like this for all the channel updates to be simple one-liner updates. All the checks around the `state = ...` lines are because there was one user where his controller returned an "unknown" client (it was missing the `is_wired` attribute in the json). This resulted in `null` devices and sites. 

I figured it was better practice to defensively guard against this possibily happening to other users in the future.

mgbowman commented:
@martinvw I finally managed to find some time to work on this

I replied to all of your outstanding comments and with the exception of the odd `INITIALIZING` check in the refresh method of the controller (I've got an idea on how to fix this) and the minor exception handling / logging issues, I believe I've addressed all the rest.

I put [a call out on the feature discussion thread](https://community.openhab.org/t/ubiquiti-unifi-binding-feature-discussion/14520/349?u=mgbowman) to help test the latest build due to the large number of changes.

Could you please review and comment on my latest code?

PS: I thought I could resolve the conflicting `pom.xml` issue with the z-way binding by commenting it out but I failed. I guess you guys will have to manually resolve that when it comes time to merge - sorry :(

martinvw commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,129 @@
+				<label>Site</label>
+				<description>The UniFi site where the client should be found (optional)</description>
+			</parameter>
+            <parameter name="considerHome" type="integer" required="false">
```
It looks like a combination tabs and spaces, please use tabs for XML files.

martinvw commented on addons/binding/org.openhab.binding.unifi/pom.xml:
```diff
@@ -0,0 +1,19 @@
+		<version>2.3.0-SNAPSHOT</version>
+	</parent>
+
+	<groupId>org.openhab.binding</groupId>
```
Optionally you can remove this line :-)

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/UniFiBindingConstants.java:
```diff
@@ -0,0 +1,49 @@
+    public static final String PARAMEMTER_CONTACT_TYPE = "contactType";
+
+    // List of all Contact Types
+    public static final String CONTACT_TYPE_NORMALLY_OPEN = "NO";
```
Is this still needed?

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,393 @@
+                throw new UniFiException("Could not communicate with the UniFi Controller");
+            } else {
+                if (logger.isTraceEnabled()) {
+                    JsonParser parser = new JsonParser();
```
You could handle this in an external method, maybe including the `if`-check

martinvw commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiController.java:
```diff
@@ -0,0 +1,393 @@
+            }
+        } catch (SocketException | SocketTimeoutException e) {
+            throw new UniFiException("Could not communicate with the UniFi Controller");
+        } catch (Exception e) {
```
Did I already ask whether it was really needed to catch all `Exception`'s?

martinvw commented:
Trying to force a rebuild

martinvw commented:
@mgbowman what indeeds seems to remain is the exception handling, thanks for your great work.

martinvw commented:
@mgbowman just to be sure, are you waiting on input from some of us, or did life just demand some extra attention again :-)

mgbowman commented:
@martinvw life unfortunately, I will get to this as soon as a I can

martinvw commented:
> Edit: What's the chances this could squeeze into 2.4?

If @kaikreuzer has time and you are able to quickly pick up the comments it could fit, but that is up to Kai

mgbowman commented:
So let me just make sure I have my procedure right here...

I should do the following...

1. Pull latest `master` from openhab2-addons into my local repo
1. Merge my `unifi-binding-v2` branch into my `unifi-binding` branch
1. Squash my commits
1. Rebase onto `unifi-binding` on to `master`
1. Force push my `unifi-binding` branch (should update this PR)

Sound about right?

martinvw commented:
Yes indeed 👍

mgbowman commented:
Done!

Will address any issues as quickly as I can hoping we can get this merged into 2.4

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerThingHandler.java:
```diff
@@ -0,0 +1,359 @@
+        UniFiDevice[] devices = controller.getDevices(site);
+        logger.debug("Found {} UniFi Device(s):", devices.length);
+        for (UniFiDevice device : devices) {
+            device.setSite(sitesCache.get(device.getSiteId()));
```
Should this not be the paramter `site` instead of the value from `sitesCache`? I'm assuming it would be the same.

mgbowman replied:
Fixed

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerThingHandler.java:
```diff
@@ -0,0 +1,359 @@
+        UniFiClient[] clients = controller.getInsights(site);
+        logger.debug("Found {} UniFi Insights(s):", clients.length);
+        for (UniFiClient client : clients) {
+            logger.debug("  {}", client);
```
These debug messages with only a space and the argument can be hard to understand if the value is unexpected. Maybe put in some context?

mgbowman replied:
I'm not sure what else to do here. I'm simply logging a list of all clients returned by the controller. The reason for the two spaces is simply to indent the output to make it a little easier to read in the console.

How would you suggest I change this?

J-N-K replied:
Maybe you could move the output to the logger above, something like „log.debug(„Found {} UniFi Insight(s): {}“, clients.length, clients);“

I think what @Hilbrand means is that there is no guarantee that there are no log lines from other modules in between and it might be hard to see that these lines belong together.

mgbowman replied:
I understand now... thanks for the clarification.

I merged the log statements and added a lazy formatter to the arrays as to not do the string concatenation unless the debug log level was active.

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiControllerRequest.java:
```diff
@@ -0,0 +1,203 @@
+    // Public API
+
+    public UniFiControllerRequest(Class<T> resultType, Gson gson, HttpClient httpClient, String host, int port) {
+        this.resultType = resultType;
```
Fixed

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiControllerRequest.java:
```diff
@@ -0,0 +1,203 @@
+@NonNullByDefault
+public class UniFiControllerRequest<T> {
+
+    private static final Charset CHARSET_UTF8 = Charset.forName("UTF-8");
```
Fixed

mgbowman commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,150 @@
+		<description>A wireless client connected to a UniFi wireless network</description>
+
+		<channels>
+		    <!--  common wired + wireless client channels  -->
```
Fixed

mgbowman commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,150 @@
+        <state readOnly="true"></state>
+    </channel-type>
+
+    <channel-type id="lastSeen">
```
Fixed

mgbowman commented:
So I see I forgot the Signed-off-by in the last 4 commits 🤦‍♂️ 

What's the recommended way to resolve this?

martinvw commented:
Using interactive rebase you can reword multiple commits

mgbowman commented:
Do I have to force push?

martinvw commented:
Yes always if you change history

mgbowman commented:
> Yes always if you change history

Thought so... git-fu is kinda rusty!

Should be resolved now.

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/README.md:
```diff
@@ -0,0 +1,141 @@
+
+## Supported Things
+
+* UniFi Controller 
```
Please name their resp. Thing type IDs

mgbowman replied:
Fixed

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java:
```diff
@@ -0,0 +1,137 @@
+        this.port = port;
+        this.username = username;
+        this.password = password;
+        this.httpClient = new HttpClient(new SslContextFactory(true)); // mgb: true = trust all
```
You should not instantiate your own HttpClient, but rather use the `HttpClientFactory` from ESH. You'll have to get it injected in your thing handler factory and pass it through the constructor of the handler down to here where it's needed.

mgbowman replied:
IIRC the issue with using `HttpClientFactory` was I couldn't use a custom `SslContext` which was necessary as the UniFi controller ships with a self-signed certificate. 

If there's another way to override the SSL verification using an injected `HttpClient` instance, I can refactor.

martinvw replied:
This is the place where the [TlsCertificateProvider](https://github.com/eclipse/smarthome/pull/6281/files#diff-1719561fd7d5b8ac889a486c612005a1R27) could shine.

The SHA2 fingerprint of the certificate used by my controller is `79 F3 F9 6C 6D 7A F3 17 85 B7 9D CD 0E B2 0C 50 62 C2 4F 1B 17 56 20 29 B6 2E 5F 70 00 4D D6 12` for controller software version `5.9.29`.

If this does not work, then the alternative could be to implement a [TlsTrustManagerProvider](https://github.com/eclipse/smarthome/pull/6281/files#diff-bf7828c630dea0acfe9071f9ed62a5faR27). Which has not yet been done before :-)

For the implementation you can use 'UniFi' for `String getHostName();` because that is what unifi set as a common name for the certificate.

mgbowman replied:
I don't think the `TlsCertificateProvider` will work as I believe everybody's certificate will be different.

That being said, maybe a `TlsTrustManagerProvider` that "accepts all" for the `config.getHostname()` parameter would work.

I'm trying to implement this, but I'm unsure of the exact "wiring" I need.

Using this in `UniFiThingHandlerFactory` I can get (and pass through) the common `HttpClient`:

```
@Reference
public void setHttpClientFactory(HttpClientFactory httpClientFactory) {
    this.httpClient = httpClientFactory.getCommonHttpClient();
}
```

However I'm not sure where / how to "install" the `TlsTrustManagerProvider`. I've tried this but it doesn't seem to be right...

```
if (UniFiControllerThingHandler.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
    handler = new UniFiControllerThingHandler((Bridge) thing, httpClient);
    bundleContext.registerService(TlsTrustManagerProvider.class.getName(), handler, null);
}
```

with `UniFiControllerThingHandler` implementing `TlsTrustManagerProvider`.

mgbowman replied:
> Maybe, maybe not? That is why I gave you my SHA2 fingerprint :-)

I have three hosts running three different versions all with different fingerprints ;-)

I manage to switch to the common `HttpClient` instance as well as implement the `TlsTrustManagerProvider`

I need to fix my Jenkins instance and ask the forum to test this since it's a rather important / critical bit of the binding that changed.

kaikreuzer commented:
> If @kaikreuzer has time and you are able to quickly pick up the comments it could fit, but that is up to Kai

I hope I was quick enough - did my best ;-)

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerThingHandler.java:
```diff
@@ -0,0 +1,359 @@
+        return config.getRefresh();
+    }
+
+    private @Nullable String invokeMethod(Object obj, String name) {
```
Fixed

mgbowman commented:
I think I've fixed everything but the two instances of catching raw `Exception`s (I know, I know this is terrible practice). I'll try and fix these two issues later today.

Anything else?

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiClientThingHandler.java:
```diff
@@ -0,0 +1,299 @@
+            if (lastSeen == null) {
+                logger.warn("Could not determine if client is online: cid = {}, lastSeen = null", config.getClientID());
+            } else {
+                Calendar considerHome = (Calendar) lastSeen.clone();
```
Just another way to calculate time. Use if you like it:
```
                online = Instant.now().isBefore( //
                        Instant.ofEpochMilli(lastSeen.getTimeInMillis()) //
                                .plus(config.getConsiderHome(), ChronoUnit.SECONDS));

```

Hilbrand commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,149 @@
+	</channel-type>
+
+	<channel-type id="uptime">
+		<item-type>Number</item-type>
```
You can make this a QuantityType: `Number:Time`.
State would be set `state = new QuantityType<Time>(client.getUptime(), SmartHomeUnits.SECOND);`
(I believe you had some discussion about this on the forum, can't remember if you decided not to use it, so just mention it here)

mgbowman replied:
Wanted to focus on getting the binding merged before 2.4 so decided to push the UoM support to the next version.

It's on my list though ;-)

mgbowman commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,149 @@
+		<item-type>Number</item-type>
+		<label>Received Signal Strength Indicator</label>
+		<description>Received Signal Strength Indicator (RSSI) of the wireless client</description>
+		<state readOnly="true"></state>
```
When I initially created this binding, I couldn't figure out exactly how the controller was displaying signal strength. I will revisit this on the next version when I implement UoM support.

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/handler/UniFiControllerThingHandler.java:
```diff
@@ -0,0 +1,359 @@
+            updateStatus(OFFLINE, COMMUNICATION_ERROR, STATUS_DESCRIPTION_COMMUNICATION_ERROR);
+        } catch (UniFiInvalidCredentialsException e) {
+            updateStatus(OFFLINE, CONFIGURATION_ERROR, STATUS_DESCRIPTION_INVALID_CREDENTIALS);
+        } catch (Exception e) {
```
IIRC I left this as `Exception` because the `ScheduledExecutorService` will stop executing scheduled tasks if an exception is thrown. I'm not sure if this has since been fixed.

[Here was the previous discussion we had on this subject](https://github.com/openhab/openhab2-addons/pull/3387#discussion_r177454322)

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/api/UniFiController.java:
```diff
@@ -0,0 +1,137 @@
+        if (!httpClient.isStarted()) {
+            try {
+                httpClient.start();
+            } catch (Exception e) {
```
Fixed with the switch to the common HttpClient instance

mgbowman commented on addons/binding/org.openhab.binding.unifi/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,149 @@
+		<item-type>Number</item-type>
+		<label>Uptime</label>
+		<description>Uptime of the client (in seconds)</description>
+		<state readOnly="true"></state>
```
Will implement in a future version with UoM support.

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiClientThingConfig.java:
```diff
@@ -0,0 +1,53 @@
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * The {@link UniFiClientThingConfig} represents the
```
Fixed

mgbowman commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiControllerThingConfig.java:
```diff
@@ -0,0 +1,60 @@
+
+/**
+ * The {@link UniFiControllerThingConfig} encapsulates all the configuration options for an instance of the
+ * {@link UniFiController}.
```
Fixed

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiControllerThingHandler.java:
```diff
@@ -0,0 +1,384 @@
+            updateStatus(OFFLINE, COMMUNICATION_ERROR, STATUS_DESCRIPTION_COMMUNICATION_ERROR);
+        } catch (UniFiInvalidCredentialsException e) {
+            updateStatus(OFFLINE, CONFIGURATION_ERROR, STATUS_DESCRIPTION_INVALID_CREDENTIALS);
+        } catch (Exception e) {
```
```suggestion
        } catch (RuntimeException | UniFiException e) {
```

Hilbrand commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiClientThingHandler.java:
```diff
@@ -0,0 +1,299 @@
+        switch (channelID) {
+            // :ap
+            case CHANNEL_AP:
+                UniFiDevice device = client.getDevice();
```
In the other switch cases `client` is checked for `null`, but not here. (You could add `@NonNullByDefault` to this class to help with this. Although this might be a version 2 feature as it requires some more work). But the null check might be needed here.

mgbowman replied:
These two `getWirelessChannelState` and `getWiredChannelState` had a null check before they're called from 

https://github.com/mgbowman/openhab2-addons/blob/unifi-binding/addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/handler/UniFiClientThingHandler.java#L248-L259

I went ahead and added `@NonNull` to the parameters and removed the existing null checks.

mgbowman replied:
It's the correct way now with `@NonNullByDefault` at the class level

kaikreuzer commented on addons/binding/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/ssl/UniFiTrustManagerProvider.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public X509ExtendedTrustManager getTrustManager() {
+        return new UniFiTrustManager();
```
I guess you should not return a new instance on every call.

mgbowman replied:
Implemented this as a singleton

kaikreuzer commented:
I'm just trying to reproduce your cyclic dependency and created a Unifi thing (without actually owning one). I found this in my log:
```
2018-12-07 18:38:06.871 [ERROR] [.h.UniFiControllerThingHandler:124  ] - Unknown error while configuring the UniFi Controllerorg.openhab.binding.unifi.internal.api.UniFiException: java.net.UnknownHostException: unifi: nodename nor servname provided, or not known
	at org.openhab.binding.unifi.internal.api.UniFiControllerRequest.getContentResponse(UniFiControllerRequest.java:157)
	at org.openhab.binding.unifi.internal.api.UniFiControllerRequest.getContent(UniFiControllerRequest.java:117)
	at org.openhab.binding.unifi.internal.api.UniFiControllerRequest.execute(UniFiControllerRequest.java:102)
	at org.openhab.binding.unifi.internal.api.UniFiController.executeRequest(UniFiController.java:113)
	at org.openhab.binding.unifi.internal.api.UniFiController.login(UniFiController.java:70)
	at org.openhab.binding.unifi.internal.api.UniFiController.start(UniFiController.java:56)
	at org.openhab.binding.unifi.internal.handler.UniFiControllerThingHandler.initialize(UniFiControllerThingHandler.java:117)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
Caused by: java.net.UnknownHostException: unifi: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at org.eclipse.jetty.util.SocketAddressResolver$Async.lambda$resolve$1(SocketAddressResolver.java:167)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680)
	... 1 common frames omitted
```
Note that this must not happen - all that should happen in this case is that the Thing goes OFFLINE with a status message that it cannot reach the host. No logging (apart from debug) must appear.

kaikreuzer commented:
Similar problem, if I create a Thing and just point it to some existing IP address in my network:
``´
2018-12-07 18:45:51.279 [ERROR] [.h.UniFiControllerThingHandler:124  ] - Unknown error while configuring the UniFi Controllerorg.openhab.binding.unifi.internal.api.UniFiException: javax.net.ssl.SSLHandshakeException: General SSLEngine problem
	at org.openhab.binding.unifi.internal.api.UniFiControllerRequest.getContentResponse(UniFiControllerRequest.java:157)
	at org.openhab.binding.unifi.internal.api.UniFiControllerRequest.getContent(UniFiControllerRequest.java:117)
	at org.openhab.binding.unifi.internal.api.UniFiControllerRequest.execute(UniFiControllerRequest.java:102)
	at org.openhab.binding.unifi.internal.api.UniFiController.executeRequest(UniFiController.java:113)
	at org.openhab.binding.unifi.internal.api.UniFiController.login(UniFiController.java:70)
	at org.openhab.binding.unifi.internal.api.UniFiController.start(UniFiController.java:56)
	at org.openhab.binding.unifi.internal.handler.UniFiControllerThingHandler.initialize(UniFiControllerThingHandler.java:117)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
Caused by: javax.net.ssl.SSLHandshakeException: General SSLEngine problem
	at sun.security.ssl.Handshaker.checkThrown(Handshaker.java:1431)
	at sun.security.ssl.SSLEngineImpl.checkTaskThrown(SSLEngineImpl.java:535)
	at sun.security.ssl.SSLEngineImpl.readNetRecord(SSLEngineImpl.java:813)
	at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:781)
	at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624)
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.fill(SslConnection.java:681)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:128)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:73)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
Caused by: javax.net.ssl.SSLHandshakeException: General SSLEngine problem
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1728)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:304)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:296)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1509)
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:216)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:979)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:919)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:916)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.security.ssl.Handshaker$DelegatedTask.run(Handshaker.java:1369)
Caused by: java.security.cert.CertificateException: No subject alternative names matching IP address 192.168.0.2 found
	at sun.security.util.HostnameChecker.matchIP(HostnameChecker.java:167)
	at sun.security.util.HostnameChecker.match(HostnameChecker.java:93)
	at sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:455)
	at sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:436)
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:252)
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:136)
	at org.eclipse.smarthome.io.net.http.internal.ExtensibleTrustManager.checkServerTrusted(ExtensibleTrustManager.java:115)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1496)
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:216)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:979)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:919)
```
Here again, there must not be any error logging - instead it should be OFFLINE with a helpful message.

mgbowman commented:
@kaikreuzer I added invalid hostname error handling - thanks for that!

While not desirable, I've implemented both `ThingHandlerFactory` and `TlsTrustManagerProvider` services in the same component `UniFiThingHandlerFactory` which forces the proper startup order allowing me to use the common `HttpClient`

I think this is the best solution for now.

kaikreuzer commented:
>  I've implemented both ThingHandlerFactory and TlsTrustManagerProvider services in the same component UniFiThingHandlerFactory

Does this really work well? I am asking because in this setup, you clearly now have a circular dependency, which is not really desirable...

mgbowman commented:
> Does this really work well?

With the services implemented as separate components, I could reproduce the out of order issue consistently (which uncovered the NPE in https://github.com/eclipse/smarthome/pull/6647).

With both services implemented in the same component, the binding only tries to create thing handlers after the `ExtensibleTrustManager` had been initialized with my `TlsTrustManagerProvider` implementation.

> I am asking because in this setup, you clearly now have a circular dependency, which is not really desirable...

I agree but here's where I'm confused with the verbiage of `Component [C]` vs `Service [S]`

```
[C] UniFiThingHandlerFactory -> [S] HttpClientFactory
[S] HttpClientFactory        -> [S] ExtensibleTrustManager
[S] ExtensibleTrustManager   -> [S] TlsTrustManagerProvider (provided by [C] UniFiThingHandlerFactory)
```

I'm not familiar with the inner workings of the OSGi and it's inter-component / inter-service lifecycle but I think the `ReferencePolicy.DYNAMIC` between `ExtensibleTrustManager` and `TlsTrustManagerProvider` is what allows the framework to properly initialize everything.

mgbowman commented:
Just a follow-up, with both services implemented in the same component, the circular reference warning goes away ¯\\\_(ツ)\_/¯ 

If a put them back into two separate components, the warning reappears

mgbowman commented:
After some feedback from a few users, I've reverted back to using an internal instance of `HttpClient`. 

There are instances where the current `TlsTrustManagerProvider` interface simply won't work. Until we can come up with a better way, I would suggest keeping it this way at least for 2.4

martinvw commented:
@mgbowman did I understand that we are now testing the release which should be okay to merge for you? 

So if @kaikreuzer has no further comments this could be mergeable :-)

kaikreuzer commented:
> So if @kaikreuzer has no further comments this could be mergeable :-)

Ok for me, just wanted to give @Hilbrand a chance to state that his comments are addressed as well.

Hilbrand commented:
I'm ok with the current changes. Nothing blocking left, other things can be addressed in the next iteration of this binding.

kaikreuzer commented:
Perfect, thanks @Hilbrand!

So let's merge then - million thanks to @mgbowman for this work and we are all looking forward to further updates for 2.5 😆 .
